### PR TITLE
WIP: Feature | Introduce "Pool Idle Timeout" connection property

### LINKED ISSUE #343: Add Pool Idle Timeout to control connection pool
From an internal email thread asking about ADO.NET connection pooling, we have this feature request from an ISV:

### Is your feature request related to a problem? Please describe.
Right now, connections returned to the pool are automatically managed and closed/disposed from the pool after 4-8 minutes of idle time. This request is to be able to configurable the amount of time a connection can be idle in the pool before it is closed. "Pool Idle Timeout".

### Describe the solution you'd like
A new connection string setting, "Pool Idle Timeout" which defines the number of seconds a connection can be idle in the pool before it is automatically closed.

### Describe alternatives you've considered
Letting the 4-8 minute auto close do its thing. Using ClearPool() and ClearAllPools().

### PR DESCRIPTION
**Pull request by cheenamalhotra**
Fixes #343

### CODE REVIEW

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlConnectionStringBuilder.cs [hunk 1]
```diff
@@ -646,6 +652,21 @@ public int MinPoolSize
             }
         }
 
+        /// <include file='..\..\..\..\..\..\..\doc\snippets\Microsoft.Data.SqlClient\SqlConnectionStringBuilder.xml' path='docs/members[@name="SqlConnectionStringBuilder"]/PoolIdleTimeout/*' />
+        public int PoolIdleTimeout
+        {
+            get { return _poolIdleTimeout; }
+            set
+            {
+                if (value < 0)
```
Comment (Wraith2) [inline src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlConnectionStringBuilder.cs]: This check for <0 will mean that users can't set it to the default value of -1 to return to the default behaviour which seems counter intuitive, both removing and setting to default value should normally yield the same result I think.

Comment (cheenamalhotra) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlConnectionStringBuilder.cs]: We're only accepting values **0 to Int32.Max** as documented [here](https://github.com/dotnet/SqlClient/pull/348/files#diff-68c5ce9ab44eb1bae56a0568df742007R597).

This customized behaviour is disabled by default (hence -1) so setting to anything other than the range is essentially not allowed.

Same is with `Connection Timeout` although it's default is 15 secs. We not really obstructing default value, but anything beyond range. I was also thinking of whether I should also check for Max value here, but didn't see that at other places, but maybe we should check everywhere?

Comment (Wraith2) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlConnectionStringBuilder.cs]: > This customized behaviour is disabled by default (hence -1) so setting to anything other than the range is essentially not allowed.

Sure, but if you enable it how do you then disable it without starting again with a new connection string builder? Or is that not a suggested use case?

Comment (cheenamalhotra) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlConnectionStringBuilder.cs]: Yes you're right. I will change it to **< -1**.

Comment (Wraith2) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlConnectionStringBuilder.cs]: Is 0 a useful value to allow? I would expect that to instantly dispose of pools which is the same thing as just not using pooling.

Comment (cheenamalhotra) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlConnectionStringBuilder.cs]: The behavior is same, yes. Allowing 0 or not, doesn't seem a problem to me.

@David-Engel if you could confirm the range as well, I can update code accordingly.

**Comment by JRahnama** [comment]
Closing as this is not mergeable. Feel free to open a new PR for this feature.

**Pull request closed without merging**